src/bootstrap/bootstrap.py

```diff
@@ -753,6 +753,9 @@ def build_bootstrap(self, color):
             target_features += ["-crt-static"]
         if target_features:
             env["RUSTFLAGS"] += " -C target-feature=" + (",".join(target_features))
+        target_linker = self.get_toml("linker", build_section)
+        if target_linker is not None:
+            env["RUSTFLAGS"] += " -C linker=" + target_linker
```

Comment:
Why is this necessary? Don't we set it above on line 745?

Reply:
Thank you for looking at this pull request.

In the MacPorts project, it is important to be able to [specify the compiler and linker](https://trac.macports.org/wiki/UsingTheRightCompiler).
To accomplish this, the [MacPorts recipe](https://github.com/macports/macports-ports/blob/85118394f6742209e4daf000961fcfcffe4ef810/lang/rust/Portfile#L141) uses
`--set=target.${platform}.cc` and `--set=target.${platform}.linker`.

However, `[target.'x86_64-apple-darwin']` is not recognized in config.toml.
Whereas `[target.x86_64-apple-darwin]` works fine.

Similarly, in a [recent change](https://github.com/rust-lang/rust/commit/d1291dc8b4ac9a98ff1d286402559e4ba5d68488#diff-9e25a089f077eae8c8cefe9586cc07498e64cd950be4c954cb90e4245e5f9fc3) setting `LD_x86_64_apple_darwin` does not seem to be recognized whereas setting RUSTFLAGS does.

Plese forgive me if I am missing something obvious.

Reply:
> However, `[target.'x86_64-apple-darwin']` is not recognized in config.toml.
> Whereas `[target.x86_64-apple-darwin]` works fine.

Can you post the error you see that makes you think this isn't recognized? This is valid toml syntax and the two should be exactly the same:
```sh
$  ./configure --set target.x86_64-apple-darwin.cc=/not/a/real/compiler
$ grep 'cc = .*/not' config.toml -B6
[target.'x86_64-apple-darwin']

# C compiler to be used to compile C code. Note that the
# default value is platform specific, and if not specified it may also depend on
# what platform is crossing to what platform.
# See `src/bootstrap/cc_detect.rs` for details.
cc = '/not/a/real/compiler'
$ toml get config.toml target.x86_64-apple-darwin.cc
"/not/a/real/compiler"
```

And if we merge your change it will break a previous bug, so I don't really want to land it as-is.

> Similarly, in a [recent change](https://github.com/rust-lang/rust/commit/d1291dc8b4ac9a98ff1d286402559e4ba5d68488#diff-9e25a089f077eae8c8cefe9586cc07498e64cd950be4c954cb90e4245e5f9fc3) setting `LD_x86_64_apple_darwin` does not seem to be recognized whereas setting RUSTFLAGS does.

What do you mean by "not recognized"? What error are you seeing?

Reply:
Ok, I think I see the issue - bootstrap.py is parsing config.toml incorrectly:
```
$ git diff
diff --git a/src/bootstrap/bootstrap_test.py b/src/bootstrap/bootstrap_test.py
index 06ca3ce21b3..8dec0de03df 100644
--- a/src/bootstrap/bootstrap_test.py
+++ b/src/bootstrap/bootstrap_test.py
@@ -73,6 +73,13 @@ class ProgramOutOfDate(unittest.TestCase):
             rustc_stamp.write("2017-06-15None")
         self.assertFalse(self.build.program_out_of_date(self.rustc_stamp_path, self.key))
 
+class ParseToml(unittest.TestCase):
+    def test_parse_quotes(self):
+        rb = bootstrap.RustBuild()
+        rb.config_toml = "[build]\ncc = 'xyz'\n[target.'x86_64-apple-darwin']\ncc = '/not/a/real/compiler'"
+        self.assertEqual(rb.get_toml("cc", "build"), "xyz")
+        cc = rb.get_toml("cc", "target.x86_64-apple-darwin")
+        self.assertEqual(cc, "/not/a/real/compiler")
 
 if __name__ == '__main__':
     SUITE = unittest.TestSuite()
@@ -80,6 +87,7 @@ if __name__ == '__main__':
     SUITE.addTest(doctest.DocTestSuite(bootstrap))
     SUITE.addTests([
         TEST_LOADER.loadTestsFromTestCase(VerifyTestCase),
+        TEST_LOADER.loadTestsFromTestCase(ParseToml),
         TEST_LOADER.loadTestsFromTestCase(ProgramOutOfDate)])
 
     RUNNER = unittest.TextTestRunner(stream=sys.stdout, verbosity=2)
$ x test bootstrap
======================================================================
FAIL: test_parse_quotes (__main__.ParseToml)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jyn/src/rust/rust/src/bootstrap/bootstrap_test.py", line 82, in test_parse_quotes
    self.assertEqual(cc, "/not/a/real/compiler")
AssertionError: None != '/not/a/real/compiler'
```
Let's fix that directly instead of removing the quotes.

Reply:
> And if we merge your change it will break a previous bug, so I don't really want to land it as-is.

I added another attempt to fix the problem that should not break a previous bug.

> > Similarly, in a [recent change](https://github.com/rust-lang/rust/commit/d1291dc8b4ac9a98ff1d286402559e4ba5d68488#diff-9e25a089f077eae8c8cefe9586cc07498e64cd950be4c954cb90e4245e5f9fc3) setting `LD_x86_64_apple_darwin` does not seem to be recognized whereas setting RUSTFLAGS does.
> 
> What do you mean by "not recognized"? What error are you seeing?

Sorry if I was not clear.
I mean the linker set in `--set=target.platform.linker` is **not** used unless RUSTFLAGS is set as was done before the [recent change](https://github.com/rust-lang/rust/commit/d1291dc8b4ac9a98ff1d286402559e4ba5d68488#diff-9e25a089f077eae8c8cefe9586cc07498e64cd950be4c954cb90e4245e5f9fc3).

> Please also add the test case I posted here.

I added the test case as requested.

Reply:
> Similarly, in a [recent change](https://github.com/rust-lang/rust/commit/d1291dc8b4ac9a98ff1d286402559e4ba5d68488#diff-9e25a089f077eae8c8cefe9586cc07498e64cd950be4c954cb90e4245e5f9fc3) setting LD_x86_64_apple_darwin does not seem to be recognized whereas setting RUSTFLAGS does.

Does it work for `target.x86_64-apple-darwin`? Or is it consistently broken for both `target.x86_64-apple-darwin` and `target.'x86_64-apple-darwin'?